pkg/cmd/pulumi/package_gen_sdk.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+
+	javagen "github.com/pulumi/pulumi-java/pkg/codegen/java"
+
+	"github.com/pulumi/pulumi/pkg/v3/codegen/dotnet"
+	gogen "github.com/pulumi/pulumi/pkg/v3/codegen/go"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/nodejs"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/python"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+)
+
+func newGenSdkCommand() *cobra.Command {
+	var language string
+	var out string
+	cmd := &cobra.Command{
+		Use:   "gen-sdk <schema_source>",
+		Args:  cobra.ExactArgs(1),
+		Short: "Generate SDK(s) from a package or schema",
+		Long: `Generate SDK(s) from a package or schema.
+
+<schema_source> can be a package name, the path to a plugin binary, or the path to a schema file.`,
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			source := args[0]
+
+			pkg, err := schemaFromSchemaSource(source)
+			if err != nil {
+				return err
+			}
+			if language == "all" {
+				for _, lang := range []string{"dotnet", "go", "java", "nodejs", "python"} {
+					err := genSDK(lang, out, pkg)
+					if err != nil {
+						return err
+					}
+				}
+				return nil
+			}
+			return genSDK(language, out, pkg)
+		}),
+	}
+	cmd.Flags().StringVarP(&language, "language", "", "all",
+		"The SDK language to generate: [nodejs|python|go|dotnet|java|all]")
+	cmd.Flags().StringVarP(&out, "out", "o", "./sdk",
+		"The directory to write the SDK to")
+	return cmd
+}
+
+func genSDK(language, out string, pkg *schema.Package) error {
+	var f func(string, *schema.Package, map[string][]byte) (map[string][]byte, error)
+	switch language {
+	case "csharp", "c#", "dotnet":
```

Comment:
Why are we accepting these additional names like "csharp" and "c#"?

Reply:
We have a set of names for the folders we generate our SDKs into (`nodejs`, `dotnet`, `python` `go`, `java`) and a set of names we accept for `pulumi convert` (`csharp`, `c#`, `go` `typescript`, `python`, `java`). I'm accepting the union to avoid being confusing, but we will only emit our standard folders ([dotnet correction](https://github.com/pulumi/pulumi/blob/df558f483c05be22828b18b2eed96e49f78025d8/pkg/cmd/pulumi/package_gen_sdk.go#L74), [nodejs correction](https://github.com/pulumi/pulumi/blob/df558f483c05be22828b18b2eed96e49f78025d8/pkg/cmd/pulumi/package_gen_sdk.go#L81)).

Reply:
The main reason I asked is because we'll have to accommodate these various names when trying to select the right plugin when we move to a plugin model.

But if we're already doing it in `pulumi convert`, we'll have to be able to support these additional "aliases" there anyway, so no concern doing it here as well.